libr/include/r_cmd.h

```diff
@@ -52,59 +59,76 @@ typedef struct r_cmd_macro_t {
 	RList *macros;
 } RCmdMacro;
 
-typedef int (*RCmdCallback)(void *user, const char *cmd);
-
 typedef struct r_cmd_item_t {
 	char cmd[64];
-	char desc[128];
-	r_cmd_callback (callback);
+	RCmdCb callback;
 } RCmdItem;
 
-typedef struct r_cmd_long_item_t {
-	char cmd[64]; /* long command */
-	int cmd_len;
-	char cmd_short[32]; /* short command */
-	char desc[128];
-} RCmdLongItem;
-
 typedef struct r_cmd_alias_t {
 	int count;
 	char **keys;
 	char **values;
 	int *remote;
 } RCmdAlias;
 
+typedef enum {
+	// for old handlers that parse their own input and accept a single string
+	R_CMD_DESC_TYPE_OLDINPUT,
+	// for old handlers that are used for recursive help
+	R_CMD_DESC_TYPE_OLDHELP,
+	// for handlers that accept argc/argv
+	R_CMD_DESC_TYPE_ARGV,
+	// for middle nodes in the tree
+	R_CMD_DESC_TYPE_INNER,
+} RCmdDescType;
+
+typedef struct r_cmd_descriptor_t {
+	RCmdDescType type;
+	char *name;
+	char *help;
+	struct r_cmd_descriptor_t *parent;
+	int n_children;
+	RPVector children;
+
+	union {
+		struct {
+			RCmdCb cb;
+		} oldinput;
+		struct {
+			const char **help_msg;
+			const char **help_detail;
+			const char **help_detail2;
+			struct r_cmd_descriptor_t *sub[127];
+		} oldhelp;
+		struct {
+			RCmdArgvCb cb;
+		} argv;
```

Comment:
argv sounds like a bad name to me for this

Reply:
is it better now? otherwise, do you have other ideas for the name?